### ISSUE TITLE
Performance regession on Macbook Air (M1, 2020)

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/MoltenVK#1209]
I've bisected a perf regression of approximately ~4ms in Dota on Macbook Air (M1, 2020) at 1440x900 to 4440a64d83ff46a3fa52945f6c6ac7084644674c

I also confirmed that setting MVK_CONFIG_FAST_MATH_ENABLED=1 fixes the performance regression.

It's plausibe to me that this is the correct default for conformance, but I wanted to make it known that this was a significant perf regression on macOS.

Please let me know how you want to resolve this, I'm willing to turn on this feature in Source 2 if we want to keep MoltenVK the way it is.

**Comment by billhollings** [comment]
@danginsburg 

>Please let me know how you want to resolve this, I'm willing to turn on this feature in Source 2 if we want to keep MoltenVK the way it is.

Definitely enable this on in Source2 to suit your app needs. That way you'll be resilient to any tweaks to this setting.

------------

I'm not sure how best to set this as a default. The original setting matched Apple's default value. On the other hand, that default value does appear to occasionally break some app visuals, on some apps. 

My sense is that conformance should probably be the default option (meaning this setting should default to disabled as it is currently), and we can document that it can be traded off for performance where appropriate by enabling this feature.

**Comment by danginsburg** [comment]
> My sense is that conformance should probably be the default option (meaning this setting should default to disabled as it is currently), and we can document that it can be traded off for performance where appropriate by enabling this feature.

Well, assuming fastMathEnabled violates the precision requirements in https://khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#spirvenv-precision-operation then yeah it probably makes sense to default it off.

The drop in performance is fairly alarming though and I worry a bit that it will be hard for devs to find.  I'm also not sure if the only option should be to turn it on or off globally.  To be honest, I'd need to go digging through the SPIR-V and Vulkan specs to fully understand how the precision requirements relate to IEEE 754.  The Metal docs aren't very explicit about what this does but I assume it's something like -ffast-math in gcc.  I also wonder how it relates to https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode in SPIR-V.

**Comment by HansKristian-Work** [comment]
In SPIRV-Cross we have the option of using explicit fast:: and precise::, but it's currently only used for min/max/clamp to deal with NaN-aware versions.

One idea is that FDiv should always be fast::, a precise:: div in MSL is probably not going to be great. Similar for trancendentals.

**Comment by greg-lunarg** [comment]
@danginsburg With regards to the SPIR-V Fast Math Mode, isn't it a show-stopper for you that these are only available for Kernel (OpenCL) execution?

**Comment by greg-lunarg** [comment]
In SPIR-V there is the RelaxedPrecision decoration. In GLSL there is mediump and lowp. In HLSL there is minNfloat types.

The problem with all three of these is that the default is full precision. So unless these are used, the compiler is obligated to use full precision.

You may remember that I wrote a pass in spirv-opt which relaxes all floating point. The problem is that it is a blunt hammer and there is no convenient way to express exceptions.

It seems that in order to attempt what you are thinking about, it would be necessary to have a way in all three languages and compilers to express full precision just where it is needed, and use relax everywhere else.

So for HLSL, create a min32float type and have a compilation mode where anything not explicitly min32float is min16float. For GLSL, have a compilation mode where anything not explicitly highp is considered mediump. For SPIR-V, have a FullPrecision decoration which would prevent its precision to be dropped to Relaxed by the spirv-opt relaxation pass.

**Comment by billhollings** [comment]
**_tl;dr:_** Re-enable `fastMathEnabled` as the default, and let SPIR-V indicate when a shader needs it disabled.

-----------------------------------------

>I'd need to go digging through the SPIR-V and Vulkan specs to fully understand how the precision requirements relate to IEEE 754. The Metal docs aren't very explicit about what this does

The appropriate Metal document for this is the [Metal Shading Language Spec](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf).

I've scanned through this doc, and compared with the Vulkan and SPIR-V docs that @danginsburg [identifies](https://github.com/KhronosGroup/MoltenVK/issues/1209#issuecomment-764655914) above.

Overall, the Metal precision specs seem to be compatible with the Vulkan precision requirements. Metal's `fastMathEnabled` switch does indeed engage Clang/GCC's `-ffast-math` compilation option.

The Vulkan spec does state:

>By default, the implementation may perform optimizations on half, single, or double-precision floating-point instructions that ignore sign of a zero, or assume that arguments and results are not NaNs or infinities. If the entry point is declared with the `SignedZeroInfNanPreserve` execution mode, then NaNs, infinities, and the sign of zero must not be ignored.

That SPIR-V `SignedZeroInfNanPreserve` mode is equivalent to what Metal's `ffast-math` compilation removes (checks for NaN's, Infs, and Signed Zeros).

From my read therefore, Vulkan seems to be indicating that the equivalent of Metal's `ffast-math` compilation option is fine unless the shader indicates differently.

We can extract the `SignedZeroInfNanPreserve` mode from the SPIR-V conversion, and then selectively enable or disable `ffast-math` compilation based on that.

**_In the end, my vote would be to revert MoltenVK to have `fastMathEnable` enabled by default (as it is in Metal anyway), and rely on SPIR-V to identify any shaders that need it to be disabled._**

BTW...Metal `ffast-math` option does include the following additional statement:

>Fast: Allow algebraically equivalent transformations, such as reassociating floating-point operations that may dramatically change the floating-point results.

where the inclusion of the "dramatically change FP results" comment seems scary on the surface...but it could also be read to mean that the fast algos will revert to full-precision if the fast algo will result in values that are too different (ie. feature, not a bug).

**Comment by greg-lunarg** [comment]
This is basically what I was looking for, but it is a little blunter than I was thinking. I was looking for a decoration rather than a mode.

One question/concern is: will this solve the breakage problem we are seeing today? As far as I can tell, glslang does not generate this mode today under any condition. So it is very unlikely that any existing shader has this set, even if it needs it. And glslang will need some change to allow users to generate it.

Maybe we create a new compile option for glslang which generates the SignedZeroInfNanPreserve mode? Maybe --full-math? By default, the SignedZeroInfNanPreserve mode would not be generated.

Or should we reverse this and generate SignedZeroInfNanPreserve by default and use --fast-math to prevent it?

Would it make sense to assume SignedZeroInfNanPreserve for older versions of SPIR-V or Vulkan so as not to break existing apps?

**Comment by danginsburg** [comment]
> One question/concern is: will this solve the breakage problem we are seeing today?

I don't think we know.  @js6i changed the fastMathEnable mode but I don't know if we have further information about the specific issue with the shader to understand if the failure was due to any of several possilbiities.  @js6i is there any more information you can provide?

If @billhollings is correct and fastMathEnable "should" be conformant with Vulkan then I would agree we should default it back to on until we have a definitive case we can look at where we can deduce whether MoltenVK/Metal is still behaving in a Vulkan conformant way or if there is one of many other possilbiites going on (a SPIR-V cross issue, an MSL compiler bug, an Metal implementation bug, or actual differences in precision behavior beyond what the Vulkan spec guarantees).

**Comment by billhollings** [comment]
>Or should we reverse this

It turns out that the Vulkan spec I quoted above that indicates that an implementation **_may_** invoke the kinds of relaxations that `ffast-math` triggers, unless the shader goes to the trouble of identifying `SignedZeroInfNanPreserve`, is derived from the `VK_KHR_shader_float_controls` extension.

The Vulkan 1.1 spec without the `VK_KHR_shader_float_controls` extension states:

>By default, the implementation may perform optimizations on half, single, or double-precision floating-point instructions that ignore sign of a zero, or assume that arguments and results are not NaNs or infinities.

In other words, the spec tells the app to expect that NaN's, Infs, etc, will **_not_** be checked by default, and to explicit indicate if a shader requires it.

So...it looks like the path forward is revert to enabling `ffast-math` by default, and then add support for the `VK_KHR_shader_float_controls` extension in MoltenVK, to allow an app to disable `ffast-math` for individual shaders. An app could also force it globally via the MoltenVK config setting.

Similarly, based on the Vulkan spec language, I should think `glslang` should also probably only include `SignedZeroInfNanPreserve` if it is explicitly requested.

**Comment by greg-lunarg** [comment]
> Similarly, based on the Vulkan spec language, I should think glslang should also probably only include SignedZeroInfNanPreserve if it is explicitly requested.

OK. I will investigate different options for doing this.

**Comment by greg-lunarg** [comment]
Actually, given that users have some recourse here, I will wait for a request to add such support to glslang.

**Comment by billhollings** [comment]
>I would agree we should default it back to on until we have a definitive case we can look at where we can deduce whether MoltenVK/Metal is still behaving in a Vulkan conformant way

As a smoke test, I've run the 1,200 CTS tests mentioning "precision" in the name, with `ffast-math` either enabled or disabled. For the most part, the same 3% (+/- a couple here or there) of these tests fail (fast-math actually has slightly better results!).

We can follow up on those in more detail, but for now it doesn't appear there is a solid case that enabling `ffast-math` will trigger significant and consistent issues. 

As a result, I'll add a PR to revert back to enabling `ffast-math` by default. While I'm at it, I'll include support for the presence of `SignedZeroInfNanPreserve` in the SPIR-V. 

And I'll add it to the dev list to implement the `VK_KHR_shader_float_controls` extension properly at some point (it's mandatory for Vulkan 1.2 anyway).

**Comment by js6i** [comment]
To me it seems that the optimizations that fast math enables, by reordering operations, is incompatible with this part of Vulkan spec:
> Operations described as “correctly rounded” will return the infinitely precise result, x, rounded so as to be representable in floating-point.

If you start exploiting associativity and distributivity you're bound to deviate from the precise result of your expression, as they don't hold for floating point operations. This is important if you:
- write specialized numerical code and do tricks to keep errors in check, or
- want the same source expression to give same results in different compilation context, given the same input.

The latter seems to sometimes cause problems for games that expect the depth results to be equal in two separate passes, and with fast mode enabled, the output geometries are slightly different - hence the glitching. I don't have more details to provide now, but I also don't see what's to learn by analyzing particular shaders.

Regarding what is the correct default - it might be the right practical choice to have it enabled anyway. But I think I would prefer to tick the "more fps" box if needed than troubleshoot rendering artifacts.

**Comment by js6i** [comment]
Funnily, Metal spec claims correct rounding for fast math mode too, with a disclaimer that it may reorder and compute something else. What does it even mean to have correct rounding for an operation that could be transformed into an entirely different one? That's a promise with no intention to keep it.

**Comment by danginsburg** [comment]
@js6i The particular case of depth differing across different shaders is what SPIR-V Invariant is for (I think that HLSL precise compiles by glslang to NoContraction, which is related but not exactly the same).  I don't think there is guarantee of those being equivalent across shaders unless you specify Invariant, regardless of whether fastmath is enabled or not.  Did the shaders you look at have this specified?  It looks to me like SPIR-V cross would generate MSL with invariant if it encounters it in the SPIR-V: https://github.com/KhronosGroup/SPIRV-Cross/blob/84a41cd4883d6b1be180359bd95f41ddd8a948fd/spirv_msl.cpp#L6818

**Comment by js6i** [comment]
@danginsburg Good to know! Indeed after a quick look I don't see these flags in neither MSL nor SPIR-V. That could be the cause of the glitching I saw. I'll have to investigate that some more. It seems then that such unsafe optimizations are the (scary and underelaborated) default.

**Comment by billhollings** [comment]
>As a smoke test, I've run the 1,200 CTS tests mentioning "precision" in the name, with `ffast-math` either enabled or disabled. For the most part, the same 3% (+/- a couple here or there) of these tests fail (`fast-math` actually has slightly better results!).

Further review indicates that all of the variances in these CTS tests with `fast-math` enabled or disabled are attributed to differences in the handling of `Infs` and `NaNs` (surprisingly  disabling `fast-math` results in more failures relative to the CTS reference values than enabling `fast-math`).

**Comment by billhollings** [comment]
Based on this discussion, PR #1226 includes several updates to deal with both `ffast-math` and `invariance` (the two are independent, but both arose in this discussion):

- Enable MSL `ffast-math` compilation by default.
- Disable `ffast-math` for an individual shader if SPIR-V indicates `SignedZeroInfNanPreserve` execution mode has been specified.
- Support compiling MSL with position invariance if the position built-in indicates `Invariant` in the SPIRV.

**Comment by danginsburg** [comment]
@billhollings Looks good to me - interesting that you have to explicitly ask the MSL compiler to respect invariance.

**Comment by billhollings** [comment]
>interesting that you have to explicitly ask the MSL compiler to respect invariance

Yes. It is curious. We had a short discussion about that over on the related [SPIRV-Cross PR](https://github.com/KhronosGroup/SPIRV-Cross/pull/1606#issuecomment-769895845), but it's definitely not clear why.
